book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,697 @@
+
+3. If `fork` is `Some`
+    - try to push `block` onto that chain
+      - if successful add `fork` to `self.chains`
+    - broadcast `result` via `block.rsp_tx`
+    - return Some(block.hash) if `result.is_ok()`
+
+4. Else panic, this should be unreachable because `commit_block` is only
+   called when it's ready to be committed.
+
+### Summary
+
+- `Chain` represents the non-finalized portion of a single chain
+
+- `ChainSet` represents the non-finalized portion of all chains and all
+  unverified blocks that are waiting for context to be available.
+
+- `ChainSet::queue` handles queueing and or commiting blocks and
+  reorganizing chains (via `commit_block`) but not finalizing them
+
+- Finalized blocks are returned from `finalize` and must still be committed
+  to disk afterwards
+
+- `finalize` handles pruning queued blocks that are past the reorg limit
+
+## Committing non-finalized blocks
+
+Given the above structures for manipulating the non-finalized state new
+`non-finalized` blocks are commited in two steps. First we commit the block
+to the in memory state, then we finalize the lowest height block if it is
+past the reorg limit, finally we process any queued blocks and prune any that
+are now past the reorg limit.
+
+1. Try to commit or queue the block to the non-finalized state with
+   `chain_set.queue(block)?;`
+
+2. If the best chain is longer than the reorg limit
+    - Finalize the lowest height block in the best chain with
+     `let finalized = chain_set.finalize()?;`
+    - commit `finalized` to disk with `CommitFinalizedBlock`
+
+## Sled data structures
+[sled]: #sled
+
+Sled provides a persistent, thread-safe `BTreeMap<&[u8], &[u8]>`. Each map is
+a distinct "tree". Keys are sorted using lex order on byte strings, so
+integer values should be stored using big-endian encoding (so that the lex
+order on byte strings is the numeric ordering).
+
+We use the following Sled trees:
+
+| Tree                 |                  Keys |                              Values |
+|----------------------|-----------------------|-------------------------------------|
+| `hash_by_height`     | `BE32(height)`        | `block::Hash`                       |
+| `height_by_hash`     | `block::Hash`         | `BE32(height)`                      |
+| `block_by_height`    | `BE32(height)`        | `Block`                             |
+| `tx_by_hash`         | `transaction::Hash`   | `BE32(height) || BE32(tx_index)`    |
+| `utxo_by_outpoint`   | `OutPoint`            | `TransparentOutput`                 |
+| `sprout_nullifiers`  | `sprout::Nullifier`   | `()`                                |
+| `sapling_nullifiers` | `sapling::Nullifier`  | `()`                                |
+| `sprout_anchors`     | `sprout::tree::Root`  | `()`                                |
+| `sapling_anchors`    | `sapling::tree::Root` | `()`                                |
+
+Zcash structures are encoded using `ZcashSerialize`/`ZcashDeserialize`.
+
```

Comment:
```suggestion

**Note:** We do not store the cumulative work for the finalized chain, because we choose the best chain using the order of non-finalized chains. The partial work past the finalized tip can be used to calculate the correct order.

```